test/hpp_test.rb

```diff
@@ -1,5 +1,202 @@
+    assert_match %r[^#{@test_client.url}], redirect_uri.to_s
+
+    params = CGI.parse(redirect_uri.query)
+    processed_attributes.each do |key, value|
+      assert_equal value, params[key].first
+    end
+  end
+
+  def test_redirect_url_generation_with_direct_skin_details
+    attributes = {
+      :currency_code => 'GBP', :payment_amount => 10000, :ship_before_date => Date.parse('2015-10-26'),
+      :merchant_reference => 'Internet Order 12345', :session_validity => Time.parse('2015-10-26 10:30'),
+      :merchant_account => 'OtherMerchant', :skin_code => 'sk1nC0de', :shared_secret => @shared_secret2,
+    }
+
+    processed_attributes = {
+      'currencyCode' => 'GBP', 'paymentAmount' => '10000', 'shipBeforeDate' => '2015-10-26',
+      'merchantReference' => 'Internet Order 12345', 'sessionValidity' => '2015-10-26T10:30:00Z',
+      'merchantAccount' => 'OtherMerchant', 'skinCode' => 'sk1nC0de', 'merchantSig' => '3vYCvD4BMBWuFcotHzbkUPIZ0EF332cLKJKLyk2PRT8='
+    }
+
+    redirect_uri = URI(@test_request.redirect_url(attributes))
+    assert_match %r[^#{@test_client.url}], redirect_uri.to_s
+
+    params = CGI.parse(redirect_uri.query)
+    processed_attributes.each do |key, value|
+      assert_equal value, params[key].first
+    end
+  end
+
+  def test_payment_methods_url_generation
+    attributes = {
+      :currency_code => 'GBP', :payment_amount => 10000, :ship_before_date => Date.parse('2015-10-26'),
+      :merchant_reference => 'Internet Order 12345', :skin => :testing, :session_validity => Time.parse('2015-10-26 10:30')
+    }
+
+    processed_attributes = {
+      'currencyCode' => 'GBP', 'paymentAmount' => '10000', 'shipBeforeDate' => '2015-10-26',
+      'merchantReference' => 'Internet Order 12345', 'sessionValidity' => '2015-10-26T10:30:00Z',
+      'merchantAccount' => 'TestMerchant', 'skinCode' => '4aD37dJA', 'merchantSig' => 'wwTSfepCgntaoyolDuNoKObsN7HhvSiNxcqOZr3tV14='
+    }
+
+    payment_methods_uri = URI(@test_request.payment_methods_url(attributes))
+    assert_match %r[^#{@test_client.url(:directory)}], payment_methods_uri.to_s
+
+    params = CGI.parse(payment_methods_uri.query)
+    processed_attributes.each do |key, value|
+      assert_equal value, params[key].first
+    end
+  end
+
+  def test_redirect_signature_check
+    params = {
+      'authResult' => 'AUTHORISED', 'pspReference' => '1211992213193029',
+      'merchantReference' => 'Internet Order 12345', 'skinCode' => '4aD37dJA',
+      'merchantSig' => 'KJHViiWW07uQvL62iXbkwks3BMHjoXDLdXw0M4k63WY='
+    }
+
+    assert Adyen::HPP::Response.new(params).redirect_signature_check
+    assert Adyen::HPP::Response.new(params, @shared_secret1).redirect_signature_check # explicitly provided shared secret
+
+    refute Adyen::HPP::Response.new(params.merge('skinCode' => 'sk1nC0de')).redirect_signature_check
+    refute Adyen::HPP::Response.new(params, @shared_secret2).redirect_signature_check # wrong shared secret
```

Comment:
nitpick: here and in line 157, have you considered using variable names like `@incorrect_secret` and `@correct_secret` and dropping the comments?